executor/benchmark_test.go

```diff
@@ -0,0 +1,424 @@
+	for i := 0; i < len(m.p.types); i++ {
+		m.colData[i] = m.genColDatums(m.p.types[i], m.p.orders[i], m.p.rows, m.p.NDVs[i])
+	}
+	return m
+}
+
+type aggExecutorParameters struct {
+	ctx          sessionctx.Context
+	schema       *expression.Schema
+	child        Executor
+	aggFuncs     []*aggregation.AggFuncDesc
+	groupByItems []expression.Expression
+	concurrency  int
+}
+
+func buildHashAggExecutor(v *aggExecutorParameters) Executor {
+	sessionVars := v.ctx.GetSessionVars()
+	e := &HashAggExec{
+		baseExecutor:    newBaseExecutor(v.ctx, v.schema, "", v.child),
+		sc:              sessionVars.StmtCtx,
+		PartialAggFuncs: make([]aggfuncs.AggFunc, 0, len(v.aggFuncs)),
+		GroupByItems:    v.groupByItems,
+	}
+	if len(v.groupByItems) != 0 || aggregation.IsAllFirstRow(v.aggFuncs) {
+		e.defaultVal = nil
+	} else {
+		e.defaultVal = chunk.NewChunkWithCapacity(e.retTypes(), 1)
+	}
+	for _, aggDesc := range v.aggFuncs {
+		if aggDesc.HasDistinct {
+			e.isUnparallelExec = true
+		}
+	}
+	if finalCon, partialCon := sessionVars.HashAggFinalConcurrency, sessionVars.HashAggPartialConcurrency; finalCon <= 0 || partialCon <= 0 || finalCon == 1 && partialCon == 1 {
+		e.isUnparallelExec = true
+	}
+	partialOrdinal := 0
+	for i, aggDesc := range v.aggFuncs {
+		if e.isUnparallelExec {
+			e.PartialAggFuncs = append(e.PartialAggFuncs, aggfuncs.Build(v.ctx, aggDesc, i))
+		} else {
+			ordinal := []int{partialOrdinal}
+			partialOrdinal++
+			if aggDesc.Name == ast.AggFuncAvg {
+				ordinal = append(ordinal, partialOrdinal+1)
+				partialOrdinal++
+			}
+			partialAggDesc, finalDesc := aggDesc.Split(ordinal)
+			partialAggFunc := aggfuncs.Build(v.ctx, partialAggDesc, i)
+			finalAggFunc := aggfuncs.Build(v.ctx, finalDesc, i)
+			e.PartialAggFuncs = append(e.PartialAggFuncs, partialAggFunc)
+			e.FinalAggFuncs = append(e.FinalAggFuncs, finalAggFunc)
+			if partialAggDesc.Name == ast.AggFuncGroupConcat {
+				finalAggFunc.(interface{ SetTruncated(t *int32) }).SetTruncated(
+					partialAggFunc.(interface{ GetTruncated() *int32 }).GetTruncated(),
+				)
+			}
+		}
+		if e.defaultVal != nil {
+			value := aggDesc.GetDefaultValue()
+			e.defaultVal.AppendDatum(i, &value)
+		}
+	}
+	return e
+}
+
+func buildStreamAggExecutor(v *aggExecutorParameters) Executor {
+	e := &StreamAggExec{
+		baseExecutor: newBaseExecutor(v.ctx, v.schema, "", v.child),
+		groupChecker: newGroupChecker(v.ctx.GetSessionVars().StmtCtx, v.groupByItems),
+		aggFuncs:     make([]aggfuncs.AggFunc, 0, len(v.aggFuncs)),
+	}
+	if len(v.groupByItems) != 0 || aggregation.IsAllFirstRow(v.aggFuncs) {
+		e.defaultVal = nil
+	} else {
+		e.defaultVal = chunk.NewChunkWithCapacity(e.retTypes(), 1)
+	}
+	for i, aggDesc := range v.aggFuncs {
+		aggFunc := aggfuncs.Build(v.ctx, aggDesc, i)
+		e.aggFuncs = append(e.aggFuncs, aggFunc)
+		if e.defaultVal != nil {
+			value := aggDesc.GetDefaultValue()
+			e.defaultVal.AppendDatum(i, &value)
+		}
+	}
+
+	return e
+}
```

Comment:
Better to reuse the existing `buildHashAgg` and `buildStreamAgg` code, i.e, we should build specific physical plans in this test package to use them as parameters of `buildHashAgg` to generate executors.